### ISSUE TITLE
Updated model interface: add parameters support.

### ISSUE THREAD

**Issue by yebai** [yebai/Turing.jl#82]
The current interface pass in variables through a dictionary. This design requires us implicitly copy all items in the data dictionary to model's scope, which is not the most consistent design with Julia's function interface. A possible better design would be:

```julia
@model gdemo(x) = begin   
  s ~ InverseGamma(2,3)
  m ~ Normal(0,sqrt(s))
  for i=1:length(x)
    x[i] ~ Normal(m, sqrt(s)) 
  end
  return(s, m, x)
end

x = [2.0, 3.0]
sampler = Gibbs(HMC(..., :m), PG(..., :s))
chn     = @sample(gdemo(x), sampler)
```

The most important changes are:

1. Allow declaring parameters in model macro.
2. Support `=` in model declaration, e.g. `gdemo(x) = statement_block`. We might want to depreciate the syntax without `=` in the future (i.e. `gdemo(x) statement_block`) to stay in line with Julia's assignment form of [function](http://docs.julialang.org/en/release-0.4/manual/functions/) definitions.
3. The twiddle macro will determine the type of left-hand-side (LHS) variable through:
           -  data: LHS variable can be found in model's parameter list.
           -  parameter: otherwise.

EDIT (28 FEB 2017): removed obsolete comment.

**Comment by xukai92** [comment]
@yebai What do you mean by "Note: we need to fetch x from gdemo(; data = Data(...))."?

**Comment by yebai** [comment]
that comment is obsolete, sorry I forgot to delete it when copying the code.

**Comment by xukai92** [comment]
No worries. 

There is a previous design decision which may affect the implementation of this update. Can you have a look at Gitter for discussion?

**Comment by yebai** [comment]
Closed by PR #85
